iroh-net/src/magic_endpoint.rs

```diff
@@ -649,25 +678,130 @@ impl MagicEndpoint {
     }
 }
 
-/// Accept an incoming connection and extract the client-provided [`PublicKey`] and ALPN protocol.
-pub async fn accept_conn(
-    mut conn: quinn::Connecting,
-) -> Result<(PublicKey, String, quinn::Connection)> {
-    let alpn = get_alpn(&mut conn).await?;
-    let conn = conn.await?;
-    let peer_id = get_remote_node_id(&conn)?;
-    Ok((peer_id, alpn, conn))
+/// Future produced by [`MagicEndpoint::accept`].
+#[derive(Debug)]
+#[pin_project::pin_project]
+pub struct Accept<'a> {
+    #[pin]
+    inner: quinn::Accept<'a>,
+    magic_ep: MagicEndpoint,
 }
 
-/// Extract the ALPN protocol from the peer's TLS certificate.
-pub async fn get_alpn(connecting: &mut quinn::Connecting) -> Result<String> {
-    let data = connecting.handshake_data().await?;
-    match data.downcast::<quinn::crypto::rustls::HandshakeData>() {
-        Ok(data) => match data.protocol {
-            Some(protocol) => std::string::String::from_utf8(protocol).map_err(Into::into),
-            None => bail!("no ALPN protocol available"),
-        },
-        Err(_) => bail!("unknown handshake type"),
+impl<'a> Future for Accept<'a> {
+    type Output = Option<Connecting>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        match this.inner.poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Ready(Some(inner)) => Poll::Ready(Some(Connecting {
+                inner,
+                magic_ep: this.magic_ep.clone(),
+            })),
+        }
+    }
+}
+
+/// In-progress connection attempt future
+#[derive(Debug)]
+#[pin_project::pin_project]
+pub struct Connecting {
+    #[pin]
+    inner: quinn::Connecting,
+    magic_ep: MagicEndpoint,
+}
+
+impl Connecting {
+    /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security.
+    pub fn into_0rtt(self) -> Result<(quinn::Connection, quinn::ZeroRttAccepted), Self> {
+        match self.inner.into_0rtt() {
+            Ok((conn, zrtt_accepted)) => {
+                // If we can't notify the rtt-actor that's not great but not critical.
```

Comment:
wrong comment?

Reply:
similar, applies to the following let-ok statements